manager/assets/modext/widgets/core/modx.grid.js

```diff
@@ -495,14 +515,190 @@ Ext.extend(MODx.grid.Grid,Ext.grid.EditorGridPanel,{
+            });
+        }
+    }
+
+
+    /* User Group-Level Permissions Checks for the calling "class" object */
+
+    /**
+     * @property {Function} setUserCanEdit - Assigns a value to userCanEdit property based on
+     * the user's permissions; used to adjust which menu items are available, whether to render links
+     * to and item's editing page, and css cues across many grid classes
+     *
+     * @param {Array} permissions - A set of permissions keys to evaluate; note that many areas currently
+     * rely on a pair of permissions (save_x and edit_x), both of which must be enabled to edit a grid item
+     *
+     * @return void
+     */
+    ,setUserCanEdit: function(permissions) {
+        permissions = permissions.map(item => item.trim());
+        this.userCanEdit = permissions.every(permission => MODx.perm[permission]);
+    }
+
+    /**
+     * @property {Function} setUserCanCreate - Assigns a value to userCanCreate property based on
+     * the user's permissions; used to adjust which menu items are available (namely the Duplicate item)
+     * and whether to render the Create button in the grid's toolbar
+     *
+     * @param {Array} permissions - A set of permissions keys to evaluate; note that many areas currently
+     * rely on a pair of permissions (save_x and new_x), both of which must be enabled to create/duplicate a grid item
+     *
+     * @return void
+     */
+    ,setUserCanCreate: function(permissions) {
+        permissions = permissions.map(item => item.trim());
+        this.userCanCreate = permissions.every(permission => MODx.perm[permission]);
+    }
+
+    /**
+     * @property {Function} setUserCanDelete - Assigns a value to userCanDelete property based on
+     * the user's permissions; used to adjust which menu items are available in the context menus
+     * and whether to render the Delete menu item within a grid toolbar's Batch button
+     *
+     * @param {Array} permissions - A set of permissions keys to evaluate
+     *
+     * @return void
+     */
+    ,setUserCanDelete: function(permissions) {
+        permissions = permissions.map(item => item.trim());
+        this.userCanDelete = permissions.every(permission => MODx.perm[permission]);
+    }
+
+
+    /* Record-Level Permissions Checks, for objects with specific policies */
+
+    ,userCanEditRecord: function(record) {
+        const permissions = record.json.permissions;
+        return !Ext.isEmpty(permissions) && permissions.update === true ? true : false ;
+    }
+
+    ,userCanDeleteRecord: function(record) {
+        const permissions = record.json.permissions;
+        return !Ext.isEmpty(permissions) && !record.json.isProtected && permissions.delete === true ? true : false ;
+    }
+
+    ,userCanDuplicateRecord: function(record) {
+        const permissions = record.json.permissions;
+        return !Ext.isEmpty(permissions) && permissions.duplicate === true ? true : false ;
```

Comment:
It confuses me that `permissions` is declared as an array, but at the same time it has such properties as `update`, `delete` and `duplicate`.
The project needs TypeScript...

Reply:
Perhaps I could refine the property naming to bring clarity to their differences. 

To explain my reasoning, there are two levels of permissions to evaluate: the primary, class-level ones (whether or not you have the permission to create/edit/delete, say, a _media source_ in general); and the secondary, object-level ones (where a policy, such as _Media Source User_, is applied within a user group's access permissions to further limit access to a particular source). The majority of MODX's grids do not need to evaluate an object-level permission, as those only apply to media sources, contexts, namespaces,  resource groups, and element categories.

Looking at each of the three grid classes I updated, you'll see (using the sources one again as example) this set of statements toward the end of its config:
```
this.gridMenuActions = ['edit', 'delete', 'duplicate'];

this.setUserCanEdit(['source_save', 'source_edit']);
this.setUserCanCreate(['source_save']);
this.setUserCanDelete(['source_delete']);
```
The `gridMenuActions` array indicates the _possible_ actions the user could take on this grid. The next three methods called (`setUserCanX`) check the MODx.perm object for the action's required permissions using the passed permissions keys. The `userCanX` set in the process are globally available in the class. This provides a universal way to define and utilize these primary permissions.

I've revised how secondary permissions get passed on a row-by-row basis, via an associative array on the php/processor end which is converted to an object on the js end.  I felt it was much more clear to have
```
permissions: {
   edit: true
}
```
as opposed to what is currently used in many places
`cls: 'pedit'`

Bottom line: Where these object-level permissions checks apply, they'll only be called if the primary class-level permissions checks pass.

Reply:
Made a handful of variable name changes to further clarify.